manager/orchestrator/taskreaper/task_reaper.go

```diff
@@ -251,7 +251,12 @@ func (tr *TaskReaper) tick() {
 
 			runningTasks := 0
 			for _, t := range historicTasks {
-				if t.DesiredState <= api.TaskStateRunning || t.Status.State <= api.TaskStateRunning {
+				// Skip tasks which are desired to be running but the current state
+				// is less than or equal to running.
+				// This check is important to ignore tasks which are running or need to be running,
+				// but to delete tasks which are either past running,
+				// or have not reached running but need to be shutdown (because of a service update, for example).
+				if t.DesiredState == api.TaskStateRunning && t.Status.State <= api.TaskStateRunning {
```

Comment:
I'm not sure this is really needed. Don't new tasks always start with desired state running?

Reply:
I think so:  https://github.com/docker/swarmkit/blob/master/api/objects.proto#L230

Reply:
Not sure why you think this is not needed. We need to shutdown tasks with desired state SHUTDOWN and actual state < ASSIGNED.

This condition is to skip tasks which are not those covered by the condition above. LMK if this is not clear and we can discuss IRL. @nishanttotla

Reply:
Oh oops, I missed the change from `||` to `&&`, because previously it would count a task with desired state of `SHUTDOWN` and current state of `PENDING` to be running - thanks for fixing.

Reply:
Oops, I missed it too. This is needed, sorry for the sliip.